bundles/org.openhab.binding.chatgpt/src/main/java/org/openhab/binding/chatgpt/internal/hli/ChatGPTHLIService.java

```diff
@@ -0,0 +1,462 @@
+    private List<ChatMessage> messages = new ArrayList<>();
+
+    private LocalTime lastMessageTime = LocalTime.now();
+    private List<ChatTools> tools = new ArrayList<>();
+    private final Logger logger = LoggerFactory.getLogger(ChatGPTHLIService.class);
+    private final Map<String, ChatFunction> FUNCTIONS = new HashMap<>();
+    private @Nullable ItemRegistry itemRegistry;
+    private @Nullable EventPublisher eventPublisher;
+    private @Nullable ChatGPTConfiguration config;
+    private MetadataRegistry metadataRegistry;
+
+    @Activate
+    public ChatGPTHLIService(@Reference ItemRegistry itemRegistry, @Reference EventPublisher eventPublisher,
+            @Reference MetadataRegistry metadataRegistry) {
+        this.itemRegistry = itemRegistry;
+        this.eventPublisher = eventPublisher;
+        this.metadataRegistry = metadataRegistry;
+
+        try (InputStream is = Thread.currentThread().getContextClassLoader().getResourceAsStream("/json/tools.json");
+                InputStreamReader reader = new InputStreamReader(is, StandardCharsets.UTF_8)) {
+
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode node = mapper.readTree(reader);
+
+            try {
+                this.tools = Arrays.asList(mapper.treeToValue(node, ChatTools[].class));
+            } catch (JsonProcessingException e) {
+                logger.debug("Error processing tools.json", e);
+                this.tools = new ArrayList<>();
+            }
+
+        } catch (IOException e) {
+            logger.debug("Error reading tools.json", e);
+            this.tools = new ArrayList<>();
+        }
+
+        for (ChatTools tool : tools) {
+            logger.debug("Loaded tool: {}", tool.getFunction().getName());
+        }
+
+        FUNCTIONS.clear();
+        FUNCTIONS.putAll(tools.stream().collect(HashMap::new, (map, tool) -> {
+            ChatFunction function = tool.getFunction();
+            String functionName = function.getName();
+
+            map.put(functionName, function);
+        }, HashMap::putAll));
+
+        ChatFunction itemControlFunction = FUNCTIONS.get("items_control");
+        if (itemControlFunction != null) {
+
+            itemControlFunction.setParametersClass(ItemsControl.class);
+
+            itemControlFunction.setExecutor(p -> {
+                ItemsControl parameters = (ItemsControl) p;
+
+                return sendCommand(parameters.getName(), parameters.getState());
+            });
+        }
+
+        logger.debug("ChatGPTHLIService activated");
+    }
+
+    @Override
+    public String getId() {
+        return SERVICE_ID;
+    }
+
+    @Override
+    public Set<String> getSupportedGrammarFormats() {
+        return Set.of();
+    }
+
+    @Override
+    public String interpret(Locale locale, String text) {
+
+        String requestBody = prepareRequestBody(text);
+        if (requestBody == null) {
+            return "Failed to prepare request body";
```

Comment:
Wouldn't an exception the better choice in such a case?

Reply:
It was planned that, upon request, the interpreted result would be sent to TTS and voiced. In all cases where requestBody could be null, an error would be logged within the prepareRequestBody function itself.